src/main/kotlin/de/gleex/pltcmd/model/mapgenerators/ui/PreviewGenerationListener.kt

```diff
@@ -0,0 +1,36 @@
+package de.gleex.pltcmd.model.mapgenerators.ui
+
+import de.gleex.pltcmd.game.GameWorld
+import de.gleex.pltcmd.model.mapgenerators.MapGenerationListener
+import de.gleex.pltcmd.model.terrain.AverageTerrain
+import de.gleex.pltcmd.model.terrain.TerrainHeight
+import de.gleex.pltcmd.model.terrain.TerrainType
+import de.gleex.pltcmd.model.world.Coordinate
+import org.hexworks.zircon.api.data.Position3D
+
+/** Shows the current state of the world in a single sector sized [GameWorld] while the map is beeing generated. */
+class PreviewGenerationListener(generatedWorldWidth: Int, generatedWorldHeight: Int, private val previewWorld: IncompleteMapGameArea) : MapGenerationListener {
+    private val displaySize = previewWorld.actualSize.to2DSize()
+    private val tilesPerPreviewWidth = generatedWorldWidth / displaySize.width
+    private val tilesPerPreviewHeight = generatedWorldHeight / displaySize.height
+
+    private val averageTerrainMap: MutableMap<Position3D, AverageTerrain> = mutableMapOf()
+
+    override fun terrainGenerated(coordinate: Coordinate, terrainHeight: TerrainHeight?, terrainType: TerrainType?) {
+        val previewCoordinate = mapToPreview(coordinate)
+
+        val terrainData = averageTerrainMap.getOrPut(previewCoordinate, { AverageTerrain() })
+        terrainData.mappedTiles[coordinate] = Pair(terrainHeight, terrainType)
```

Comment:
I think it is not nessecary to make `mappedTiles` public. There is no need to fiddle around with `Pair` outside of `AverageTerrain` (it can keep its inner data sctructures for itself).
```suggestion
        terrainData.put(coordinate, terrainHeight, terrainType)
```
(I thought about a nicer syntax for this call but everything else that came to my mind wasn't more readable or easier to use)

Reply:
You're right, that the inner field should not be accessible.